File: intern/137.single-number-ii.js
/*
 * @lc app=leetcode id=137 lang=javascript
 *
 * [137] Single Number II
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @return {number}
 */
var singleNumber = function (nums) {
    let result = 0;
    for (let i = 0; i < 32; i++) {
        const mask = 0b1 << i;
        let bitCount = 0;
        for (const num of nums) {
            if ((num & mask) !== 0) {
                bitCount++;
            }
        }
        if (bitCount % 3) {
            result += mask;
        }
    }
    return result;
};
// @lc code=end

File: prev-2024/137.single-number-ii.cpp
/*
 * @lc app=leetcode id=137 lang=cpp
 *
 * [137] Single Number II
 */
#include <vector>

using std::vector;

// @lc code=start
class Solution {
 public:
  int singleNumber(vector<int>& nums) {
    int once = 0;
    int twice = 0;
    for (auto& num : nums) {
      // `num`s appears once so far
      once = (once ^ num) & ~twice;
      // `num`s appears twice so far
      twice = (twice ^ num) & ~once;
    }
    return once;
  }
};
// @lc code=end
